common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -194,122 +215,143 @@ private IRequestTelemetryCache createLastRequestTelemetryCache(@NonNull final Co
-                !TextUtils.isEmpty(lastTelemetry.getCommonTelemetry().get(Schema.Key.API_ID));
+    @Nullable
+    private String getErrorFromCommandResult(final CommandResult commandResult) {
+        if (commandResult.getStatus() == CommandResult.ResultStatus.ERROR) {
+            final BaseException baseException = (BaseException) commandResult.getResult();
+            return baseException.getErrorCode();
+        } else if (commandResult.getStatus() == CommandResult.ResultStatus.CANCEL) {
+            return "user_cancel";
+        } else {
+            return null;
+        }
+    }
+
+    private boolean isTelemetryLoggedByServer(@NonNull final BaseCommand command, @NonNull final CommandResult commandResult) {
+        // This was a local operation - we didn't reach token endpoint and hence telemetry wasn't sent
+        if (!(command instanceof TokenCommand)) {
+            return false;
+        }
+
+        if (commandResult.getStatus() == CommandResult.ResultStatus.ERROR) {
+            BaseException baseException = (BaseException) commandResult.getResult();
+            if (!(baseException instanceof ServiceException)) {
+                // Telemetry not logged by server as the exception is a local exception
+                // (request did not reach token endpoint)
+                return false;
+            } else {
+                final ServiceException serviceException = (ServiceException) baseException;
+                final int statusCode = serviceException.getHttpStatusCode();
+                // for these status codes, headers aren't logged by ests
+                return !(statusCode == ServiceException.DEFAULT_STATUS_CODE ||
+                        statusCode == 429 ||
+                        statusCode >= 500);
+            }
+        } else if (commandResult.getStatus() == CommandResult.ResultStatus.CANCEL) {
+            // we did not go to token endpoint
+            return false;
+        } else if (commandResult.getStatus() == CommandResult.ResultStatus.COMPLETED) {
+            if (commandResult.getResult() instanceof ILocalAuthenticationResult) {
+                final ILocalAuthenticationResult localAuthenticationResult = (ILocalAuthenticationResult) commandResult.getResult();
+                if (localAuthenticationResult.isServicedFromCache()) {
+                    // we did not go to token endpoint
+                    return false;
+                }
+            } else {
+                // command probably wasn't a token command - we should never get here in that case
+                return false;
+            }
+        }
+
+        // if we get here that means we went to token endpoint and headers were logged by sts
+        return true;
     }
 
-    String getCurrentTelemetryHeaderString() {
+    private int getLoggedTelemetryIndex() {
+        return ((SharedPreferencesLastRequestTelemetryCache) mLastRequestTelemetryCache)
+                .getLastTelemetryIndexSentInHeaderFromCache();
```

Comment:
Shouldn't this value be part of the lastRequestTelemetry object (like the associated failedRequest/error arrays).

Reply:
No, this value isn't actually part of telemetry i.e. it isn't actually a field in the schema and hence I didn't add it LastRequest Telemetry class.